redact/redact_test.go

```diff
@@ -237,3 +238,104 @@ func TestJSON(t *testing.T) {
+	multiRedact := []*Redact{
+		newTestRedact(t, "foobar", "baz"),
+		newTestRedact(t, "baz", "<REDACTED>"),
+	}
+
+	tcs := []struct {
+		name   string
+		input  [][]*Redact
+		expect []*Redact
+	}{
+		// Nil slice alone, first, last, middle
+		{
+			name:   "Flatten should return empty redact slice for nil slice input",
+			input:  [][]*Redact{nilSlice},
+			expect: make([]*Redact, 0),
+		},
+		{
+			name:   "Flatten should treat a nil slice (first arg) correctly",
+			input:  [][]*Redact{nilSlice, singleRedact},
+			expect: singleRedact,
+		},
+		{
+			name:   "Flatten should treat mixed args (multiRedact, singleRedact, nil slice) correctly",
+			input:  [][]*Redact{multiRedact, singleRedact, nilSlice},
+			expect: []*Redact{multiRedact[0], multiRedact[1], singleRedact[0]},
+		},
+		{
+			name:   "Flatten should treat mixed args (multiRedact, nil slice, singleRedact) correctly",
+			input:  [][]*Redact{multiRedact, nilSlice, singleRedact},
+			expect: []*Redact{multiRedact[0], multiRedact[1], singleRedact[0]},
+		},
+		// Single arg
+		{
+			name:   "Flatten should treat a single redact input correctly",
+			input:  [][]*Redact{singleRedact},
+			expect: singleRedact,
+		},
+		// Multi-arg
+		{
+			name:   "Flatten should treat a multi-redact slice input correctly",
+			input:  [][]*Redact{multiRedact},
+			expect: multiRedact,
+		},
+		{
+			name:   "Flatten should treat mixed-length inputs correctly (1)",
+			input:  [][]*Redact{multiRedact, singleRedact},
+			expect: []*Redact{multiRedact[0], multiRedact[1], singleRedact[0]},
+		},
+		{
+			name:   "Flatten should treat mixed-length inputs correctly (2)",
+			input:  [][]*Redact{singleRedact, multiRedact},
+			expect: []*Redact{singleRedact[0], multiRedact[0], multiRedact[1]},
+		},
+		// empty slice alone, first, last, middle
+		{
+			name:   "Flatten should return empty redact slice for empty redact slice input",
+			input:  [][]*Redact{emptySlice},
+			expect: make([]*Redact, 0),
+		},
+		{
+			name:   "Flatten should treat an empty slice (first arg) correctly",
+			input:  [][]*Redact{emptySlice, singleRedact},
+			expect: singleRedact,
+		},
+		{
+			name:   "Flatten should treat mixed args (multiRedact, singleRedact, empty slice) correctly",
+			input:  [][]*Redact{multiRedact, singleRedact, emptySlice},
+			expect: []*Redact{multiRedact[0], multiRedact[1], singleRedact[0]},
+		},
+		{
+			name:   "Flatten should treat mixed args (multiRedact, empty slice, singleRedact) correctly",
+			input:  [][]*Redact{multiRedact, emptySlice, singleRedact},
+			expect: []*Redact{multiRedact[0], multiRedact[1], singleRedact[0]},
+		},
+		{
+			name:   "Flatten should treat mixed args (nil slice, multiRedact, empty slice, singleRedact) correctly",
+			input:  [][]*Redact{nilSlice, multiRedact, emptySlice, singleRedact},
+			expect: []*Redact{multiRedact[0], multiRedact[1], singleRedact[0]},
+		},
+	}
+	// Run assertions
+	for _, tc := range tcs {
```

Comment:
I see that this is the same pattern used in other tests in this file, but I'd propose a slight change - at least for consideration!

If we change lines 329-332 to this, instead:

```go
	for _, tc := range tcs {
		t.Run(tc.name, func(t *testing.T) {
			result := Flatten(tc.input...)
			assert.Equal(t, tc.expect, result, tc.name)
		})
	}
```

then, instead of having one test that loops through 13 cases, we wind up with 13 subtests that are part of one bigger test. That means output like this:

```
go test ./redact -run TestFlatten -v                                                                                                                                                                                      [12:20:19]
=== RUN   TestFlatten
--- PASS: TestFlatten (0.00s)
PASS
ok      github.com/hashicorp/hcdiag/redact      0.976s
```

becomes output like this:

```
go test ./redact -run TestFlatten -v                                                                                                                                                                                      [12:20:15]
=== RUN   TestFlatten
=== RUN   TestFlatten/Flatten_should_return_empty_redact_slice_for_nil_slice_input
=== RUN   TestFlatten/Flatten_should_treat_a_nil_slice_(first_arg)_correctly
=== RUN   TestFlatten/Flatten_should_treat_mixed_args_(multiRedact,_singleRedact,_nil_slice)_correctly
=== RUN   TestFlatten/Flatten_should_treat_mixed_args_(multiRedact,_nil_slice,_singleRedact)_correctly
=== RUN   TestFlatten/Flatten_should_treat_a_single_redact_input_correctly
=== RUN   TestFlatten/Flatten_should_treat_a_multi-redact_slice_input_correctly
=== RUN   TestFlatten/Flatten_should_treat_mixed-length_inputs_correctly_(1)
=== RUN   TestFlatten/Flatten_should_treat_mixed-length_inputs_correctly_(2)
=== RUN   TestFlatten/Flatten_should_return_empty_redact_slice_for_empty_redact_slice_input
=== RUN   TestFlatten/Flatten_should_treat_an_empty_slice_(first_arg)_correctly
=== RUN   TestFlatten/Flatten_should_treat_mixed_args_(multiRedact,_singleRedact,_empty_slice)_correctly
=== RUN   TestFlatten/Flatten_should_treat_mixed_args_(multiRedact,_empty_slice,_singleRedact)_correctly
=== RUN   TestFlatten/Flatten_should_treat_mixed_args_(nil_slice,_multiRedact,_empty_slice,_singleRedact)_correctly
--- PASS: TestFlatten (0.00s)
    --- PASS: TestFlatten/Flatten_should_return_empty_redact_slice_for_nil_slice_input (0.00s)
    --- PASS: TestFlatten/Flatten_should_treat_a_nil_slice_(first_arg)_correctly (0.00s)
    --- PASS: TestFlatten/Flatten_should_treat_mixed_args_(multiRedact,_singleRedact,_nil_slice)_correctly (0.00s)
    --- PASS: TestFlatten/Flatten_should_treat_mixed_args_(multiRedact,_nil_slice,_singleRedact)_correctly (0.00s)
    --- PASS: TestFlatten/Flatten_should_treat_a_single_redact_input_correctly (0.00s)
    --- PASS: TestFlatten/Flatten_should_treat_a_multi-redact_slice_input_correctly (0.00s)
    --- PASS: TestFlatten/Flatten_should_treat_mixed-length_inputs_correctly_(1) (0.00s)
    --- PASS: TestFlatten/Flatten_should_treat_mixed-length_inputs_correctly_(2) (0.00s)
    --- PASS: TestFlatten/Flatten_should_return_empty_redact_slice_for_empty_redact_slice_input (0.00s)
    --- PASS: TestFlatten/Flatten_should_treat_an_empty_slice_(first_arg)_correctly (0.00s)
    --- PASS: TestFlatten/Flatten_should_treat_mixed_args_(multiRedact,_singleRedact,_empty_slice)_correctly (0.00s)
    --- PASS: TestFlatten/Flatten_should_treat_mixed_args_(multiRedact,_empty_slice,_singleRedact)_correctly (0.00s)
    --- PASS: TestFlatten/Flatten_should_treat_mixed_args_(nil_slice,_multiRedact,_empty_slice,_singleRedact)_correctly (0.00s)
PASS
ok      github.com/hashicorp/hcdiag/redact      0.146s
```

In the case of failures, having the test name at the top of the output can be handy. Instead of this:

```
--- FAIL: TestFlatten (0.00s)
    redact_test.go:337: 
                Error Trace:    redact_test.go:337
                Error:          Not equal: 
```

we can get this:

```
--- FAIL: TestFlatten (0.00s)
    --- FAIL: TestFlatten/Flatten_should_treat_mixed_args_(multiRedact,_empty_slice,_singleRedact)_correctly (0.00s)
        redact_test.go:332: 
                Error Trace:    redact_test.go:332
                Error:          Not equal: 
 ...
```

Additionally, it allows running individual subtests, such as:

```shell
go test ./redact -run TestFlatten/"Flatten should treat mixed args \(multiRedact, empty slice, singleRedact\) correctly" -v
# the string is a regex so we have to escape the parentheses in this particular case
```

Reply:
That's so much better, and thank you for the detailed explanation! Adding this.